### PR TITLE
resize on append

### PR DESCRIPTION
**Pull request by MatthewBeaudouinLafon**
Added resizing to append. Also checks to make sure that `max_size` doesn't get larger than the maximum int size. 

Also, we should consider changing `size` and `max_size` to an unsigned bit to double our maximum array length.

Diff looks bad because I fixed some comment spacing.

### CODE REVIEW

File: src/shvec.c [hunk 1]
```diff
@@ -3,16 +3,17 @@
 #include <stdlib.h>
 
 typedef struct {
-    int size;        // number of initialized values
-    int max_size;    // max number of values
-    int *data;   // array of values
+    int size;       // number of initialized values
```
Comment (labseven) [inline src/shvec.c:9]: Nice

File: src/shvec.c [hunk 2]
```diff
@@ -36,18 +37,39 @@ int shvec_create(){
...
-* Appends a value to the shvec, reallocates the array if necessary
+ * Appends a value to the shvec, reallocates the array if necessary
  */
 int shvec_append(int id, int value){
     if(shvec_array[id].size < shvec_array[id].max_size){
         shvec_array[id].size++;
         shvec_array[id].data[shvec_array[id].size] = value;
         return 0;
     } else {
+        int* old_data_ptr = shvec_array[id].data;
```
Comment (labseven) [inline src/shvec.c]: Can we split this off into a function shvec_expand(int id)?

File: src/shvec.c [hunk 3]
```diff
@@ -36,18 +37,39 @@ int shvec_create(){
...
+        int old_size = shvec_array[id].size;
+
         // realloc and update max_size
-        // todo (Matt)
-        fprintf(stderr, "ERROR: could not resize array\n");
-        return 1;
+        while (shvec_array[id].size < shvec_array[id].max_size) {
```
Comment (labseven) [inline src/shvec.c]: Use global GROWTH_FACTOR for new max_size

File: src/shvec.c [hunk 4]
```diff
@@ -36,18 +37,39 @@ int shvec_create(){
...
+            shvec_array[id].max_size = shvec_array[id].max_size << 1;
+            
+            if (shvec_array[id].max_size >= SHVEC_SIZE_CAP) {
```
Comment (labseven) [inline src/shvec.c]: Will this actually catch an overflow? if you overflow, max_size will become negative...

**Review by labseven — CHANGES_REQUESTED** [review]
okay once the two comments are adressed

File: src/shvec.c (hunk 4 above)
Comment (labseven) [reply src/shvec.c]: oh, I see, SHVEC_SIZE_CAP is one bit smaller than the largest int

File: src/shvec.c (hunk 2 above)
Comment (SeunginLyu) [inline src/shvec.c]:
```suggestion
        int* old_data = shvec_array[id].data;
```
Let's keep it consistent with the .data_ptr -> .data change

File: src/shvec.c [hunk 5]
```diff
@@ -36,18 +38,38 @@ int shvec_create(){
 }
 
 /*
-* Appends a value to the shvec, reallocates the array if necessary
+ * Expand memory allocation of shvec's data.
+ */
+int shvec_expand(int id) {
+    // Catch int overflow before changing max_size
+    while (shvec_array[id].size < shvec_array[id].max_size) {
```
Comment (labseven) [inline src/shvec.c]: either move line 52 into the while or get rid of the while.

This way, the while is redundant AND can have the program get stuck in an infinite loop

File: src/shvec.h [hunk 6]
```diff
@@ -1,8 +1,10 @@
+// cap max_size to the maximum size of a power of 2 int
 #define MAX_SHVECS 128
-#define INIT_SIZE 10
+#define INIT_SIZE (1 << 3)
 #define GROWTH_FACTOR 2
 
 int shvec_create();
+int shvec_expand(int id);
```
Comment (labseven) [inline src/shvec.h]: I would not put this into the header, nobody outside should be calling this

**Review by labseven — CHANGES_REQUESTED** [review]

**Review by labseven — APPROVED** [review]
Looks good

**Pull request merged**